Annotator add basic Tools & Panels

maciejdrozdowski opened the pull request:
0) Create dedicated annotator.types file for all internal types used in annotator

1) Basic Hooks changes
- remove annotator.hooks.tsx and move its content to dedicated simpler hooks
- Create Info Folder for all hooks related with user decisions, selected things, filters (etc etc) - all related with react state
- Create useChoices, useFetch, useInfo, useCursor, useFilter into Info folder

2) Paper Hooks changes
- remove painter hook from annotator.hooks.tsx
- move useCanvas hook into Paper folder
- Create Paper folder for ALL hooks & methods & components which use paper.js scope objects ( mainly references )
- Create Paper/Tools subfolder for all hooks related with Paper.Tool ( object which handle mouse & keyboard in various useCases )
- add 'useTools' hook which gather/manage all Tools states
- Add 'BBox' and 'Eraser' Tools for Paper/tools ( create rectangle shape + rubber )
- Create Paper/Utils subfolder for all reusable paper.js methods
- Create Paper/Part subfolder for all components which use paper.js objects
- Create Paper/Various subfolder for every other paper.js hooks

3) Add Right Panel CategoryCard & AnnotationCard - it handle enabled actions correctly

4) Add Category filter field

5) Adjust BBox & Eraser Tools Panels

maciejdrozdowski committed:
Annotator add basic Tools & Panels

0) Create dedicated annotator.types file for all internal types used in annotator

1) Basic Hooks changes
- remove annotator.hooks.tsx and move its content to dedicated simpler hooks
- Create Info Folder for all hooks related with user decisions, selected things, filters (etc etc) - all related with react state
- Create useChoices, useFetch, useInfo, useCursor, useFilter into Info folder

2) Paper Hooks changes
- remove painter hook from annotator.hooks.tsx
- move useCanvas hook into Paper folder
- Create Paper folder for ALL hooks & methods & components which use paper.js scope objects ( mainly references )
- Create Paper/Tools subfolder for all hooks related with Paper.Tool ( object which handle mouse & keyboard in various useCases )
- add 'useTools' hook which gather/manage all Tools states
- Add 'BBox' and 'Eraser' Tools for Paper/tools ( create rectangle shape + rubber )
- Create Paper/Utils subfolder for all reusable paper.js methods
- Create Paper/Part subfolder for all components which use paper.js objects
- Create Paper/Various subfolder for every other paper.js hooks

3) Add Right Panel CategoryCard & AnnotationCard - it handle enabled actions correctly

4) Add Category filter field

5) Adjust BBox & Eraser Tools Panels

wojciech-bilicki commented on client/src/Annotator/Annotator.tsx:
```diff
@@ -1,55 +1,62 @@
 
 import * as Menu from './Menu';
 import * as Panel from './Panel';
+import { Part } from './Paper';
```
you import twice from the same module. Check line 11

wojciech-bilicki commented on client/src/Annotator/Annotator.tsx:
```diff
@@ -94,41 +102,55 @@ const Annotator: React.FC<{ imageId: number }> = ({ imageId }) => {
                     {categories.length > 0 ? (
                         segmentOn ? (
-                            categoryData.filteredList.map(cat => (
+                            info.data.map(g => (
```
`g` isn't very informative variable name

wojciech-bilicki commented on client/src/Annotator/Info/useChoices.tsx:
```diff
@@ -0,0 +1,53 @@
+    const setSelected = useCallback((catId: number, annotationId?: number) => {
+        _setSelected({
+            category: catId,
+            annotation: annotationId ? annotationId : null,
```
it's annotationId, not the whole annotation object. The name of the entry in state is misleading.

wojciech-bilicki commented on client/src/Annotator/Info/useCursor.tsx:
```diff
@@ -0,0 +1,52 @@
+        let cursor: string = 'default';
+        switch (activeTool) {
+            case Tool.SELECT:
+                cursor = 'pointer';
```
values that are assigned to cursor should also be grouped in an enum since they are a closed set.

wojciech-bilicki commented on client/src/Annotator/Info/useFetch.tsx:
```diff
@@ -0,0 +1,72 @@
+import useGlobalContext from '../../common/hooks/useGlobalContext';
+import { addProcess, removeProcess } from '../../common/utils/globalActions';
+
+const useFetch = (imageId: number) => {
```
useFetch isn't a proper name for this hook. It is named like a hook that would use standard `fetch` functionality from a browser to perform XHR request.

wojciech-bilicki commented on client/src/Annotator/Info/useFilter.tsx:
```diff
@@ -0,0 +1,38 @@
+            return categories.map(o => o.id);
+        }
+        const search = searchText.toLowerCase();
+        const cats = categories.filter(category =>
```
![](http://becomeawag.com/wp-content/uploads/2019/09/coughing-cat-know-your-meme-6.png)

Please change the variable name

wojciech-bilicki commented on client/src/Annotator/Info/useInfo.tsx:
```diff
@@ -0,0 +1,121 @@
+}
+interface UseCategoriesResponse {
+    data: CategoryInfo[];
+    setCategoryEnabled: (catId: number) => void;
```
`categoryId`

wojciech-bilicki commented on client/src/Annotator/Paper/Part/AnnotationInfo.tsx:
```diff
@@ -0,0 +1,85 @@
+        const group = groupsRef.current.find(
+            o => o.data.categoryId === categoryId,
+        );
+
```
Maybe do an early return instead of nested ifs?

wojciech-bilicki commented on client/src/Annotator/Paper/Part/AnnotationInfo.tsx:
```diff
@@ -0,0 +1,85 @@
+            o => o.data.categoryId === categoryId,
+        );
+
+        if (group != null) {
```
ditto

wojciech-bilicki commented on client/src/Annotator/Paper/Part/AnnotationInfo.tsx:
```diff
@@ -0,0 +1,85 @@
+            o => o.data.categoryId === categoryId,
+        );
+
+        if (group != null) {
```
ditto

wojciech-bilicki commented on client/src/Annotator/Paper/Part/AnnotationInfo.tsx:
```diff
@@ -0,0 +1,85 @@
+        }
+    }, [id, categoryId, isSelected, groupsRef]);
+
+    return null;
```
all those functions seem a bit the same. Anything we can do to factor out a common function to make it DRY?

wojciech-bilicki commented on client/src/Annotator/Paper/Tool/useEraser.tsx:
```diff
@@ -0,0 +1,162 @@
+        strokeWidth: 1,
+        radius: 30,
+    });
+    const [radius, _setRadius] = useState(30);
```
magic repeated number

wojciech-bilicki commented on client/src/Annotator/Paper/Utils/pathUtils.tsx:
```diff
@@ -0,0 +1,95 @@
+
+    let center = new paper.Point(a.width / 2, a.height / 2);
+    let children: Array<paper.Path> = [];
+    a.segmentation.forEach(polygon => {
```
one letter variable name aren't good. I assume that your IDE has autocompletion so a longer name not gonna hurt.

wojciech-bilicki commented on client/src/Annotator/annotator.types.tsx:
```diff
@@ -0,0 +1,24 @@
+import paper from 'paper';
+
+export type Maybe<T> = T | null;
```
`| undefined` is a proper definition of `Maybe` type for javascript

maciejdrozdowski committed:
PR review fixes

- Adjust Maybe type with | undefined value - adjust all references
- Adjust valid property names for SelectedState categoryId & annotationId ( + move selectedState to types )
- Add & use cursor enum type
- Rename useFetch to useDataset
- Remove duplicate code for finding Path Object in AnnotationInfo

wojciech-bilicki commented on client/src/Annotator/Info/useChoices.tsx:
```diff
@@ -0,0 +1,49 @@
+    const setSelected = useCallback((catId: number, annotationId?: number) => {
+        _setSelected({
+            categoryId: catId,
+            annotationId: annotationId ? annotationId : null,
```
this can be changed to `annotationId || null`

wojciech-bilicki commented on client/src/Annotator/Info/useFilter.tsx:
```diff
@@ -0,0 +1,37 @@
+    }, [categories, searchText]);
+
+    const setSearchText = useCallback((value: string) => {
+        _setSearchText(value ? value : '');
```
`value || ''`

maciejdrozdowski committed:
PR review fixes part 2
